sklearn/cluster/k_means_.py

```diff
@@ -391,8 +394,6 @@ def _kmeans_single_elkan(X, n_clusters, max_iter=300, init='k-means++',
                          verbose=False, x_squared_norms=None,
                          random_state=None, tol=1e-4,
                          precompute_distances=True):
-    if sp.issparse(X):
-        raise ValueError("algorithm='elkan' not supported for sparse input X")
     X = check_array(X, order="C")
```

Comment:
Could you mentioned specifically:

```python
X = check_array(X, accept_sparse=False, order="C")
```

Reply:
You can drop this check_array altogether. It is a waste of time, unless I'm much mistaken.

And I think it would be fine to change the ValueError to a more appropriate TypeError; but the message is more helpful.